ZGEMM optimizations + some bug fixes

nakajee opened the pull request:
Lint and Tensile Client Test are passed.

nakajee commented:
* DirectToVgpr support for ZGEMM
* DirectToLds support for ZGEMM
* StoreCInUnroll support for ZGEMM
* MIArchVgpr support for ZGEMM
* increased AssertSummationElementMultiple up to 64
* fixed mismatch issue with PGR=2 + 1LDSB=0 + K=DepthU*2
* fixed incorrect waitcnt number with SCIU=1 + PGR=2 + DTVA=1
* fixed incorrect waitcnt number with DTVA=1 and non-square macroTile size (such as 128x64)
* fixed mismatch issue with DTL=1 and VW=1 and GRVW=2
* added restricton for DLT + NLC>2 + DGEMM

zaliu commented:
@imcarsonliao Would you please get someone on your team to review this PR?  Thanks very much.

rosenrodt commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -6647,43 +6663,101 @@ def mfmaIter(self, kernel, m, innerUnroll, vregSetIdx, lastKinloop=False, tail=F
             ccInsts = [None]*3
             accImOffset = self.AccVgprImagNumOffset(kernel)
+            # for firstIter, need to use accStartSrc for img instead of adding accImOffset
+            accStartSrcImg1 = accStartSrc1 if firstIter else accStartSrc1+accImOffset
```
What are the considerations for needing this special `firstIter == true` case?

nakajee replied:
This is to optimize accreg initialization.
Instead of initializing all accregs, only 8 accregs are initialized.
At the first iteration, the initialized 8 accregs are used as source accreg.

rosenrodt reviewed (commented):
Aside from some inline comments/questions, please add a few yaml files in unit tests. I have already found out a case that ZGEMM failed. But I have no way to tell whether the parameters are valid though.

<details open>
<summary>ZGEMM MFMA yaml failed case</summary>
<br>

```yaml
GlobalParameters:
  NumElementsToValidate: 256

BenchmarkProblems:
  ########################################
  # NT - standard
  ########################################
  -
    - # ProblemType
      OperationType: GEMM
      DataType: z
      TransposeA: False
      TransposeB: True
      UseBeta: True
      Batched: True

    - # BenchmarkProblemSizeGroup - Standard
      InitialSolutionParameters:
      BenchmarkCommonParameters:
        - KernelLanguage: ["Assembly"]
        - EdgeType: ["ShiftPtr"]
      ForkParameters:
        - MatrixInstruction:
          - [16, 16, 4, 1]
        - ThreadTile:
          - [ 1, 32 ]
        - WorkGroup:
          - [ 64, 4, 1 ]
        - WorkGroupMapping: [16]
        - PrefetchGlobalRead: [True]
        - PrefetchLocalRead: [True]
        - DepthU: [8]
        - StoreCInUnroll: [1]
        - PersistentKernel: [1]
        - PrefetchAcrossPersistent: [1]
        - PrefetchAcrossPersistentMode: [1]
        - AssertSummationElementMultiple: [16]
        - ScheduleIterAlg: [3]
      BenchmarkForkParameters:
      JoinParameters:
      BenchmarkJoinParameters:
      BenchmarkFinalParameters:
        - ProblemSizes:
          - Exact: [ 1024, 1024, 1, 1024 ]
```

```log
Index:  Device | Reference
[0]  elem=0 idx=0: (-6.77969e+06,-2.69644e+06)!=(-6.78006e+06,-2.69624e+06)
[1]  elem=4099 idx=4099: (2.32442e+07,-1.11045e+07)!=(2.41451e+07,5.28322e+07)
[2]  elem=8198 idx=8198: (0,0)!=(-3.09174e+06,-1.96237e+07)
[3]  elem=12297 idx=12297: (0,0)!=(7.55016e+07,-3.11986e+06)
run,problem-progress,solution-progress,operation,problem-sizes,solution,validation,time-us,gflops,empty,total-gran,tiles-per-cu,num-cus,tile0-gran,tile1-gran,cu-gran,wave-gran,mem-read-bytes,mem-write-bytes,temp-edge,clock-sys,clock-soc,clock-mem,fan-rpm,hardware-samples,enqueue-time
0,0/0,0/0,Contraction_l_Ailk_Bjlk_Cijk_Dijk,"(1024,1024,1,1024)",Cijk_Ailk_Bjlk_ZB_MT64x32x8_MI16x16x4x1_SN_K1,FAILED,913.042,2352.01,,0.930909,4.65455,110,1,1,0.930909,1,822083584,16777216,,,,,,,2022-02-08 17:57:37.271987
+ ERR2=1
```

</details>

nakajee commented on Tensile/KernelWriterAssembly.py:
```diff
@@ -6647,43 +6663,101 @@ def mfmaIter(self, kernel, m, innerUnroll, vregSetIdx, lastKinloop=False, tail=F
+              # set up offset and number of tmp vgpr to check out
+              numVgpr = numRegistersOut
+              if not forceGenerate:
+                if ai not in zgemmVaddSrcCheck[arryaIndex]:
```
Updated.

nakajee commented:
Regarding the fail you reported, it hit some issues with DepthU=8.
I did not test DepthU=8 since it was not good for performance.
I fixed multiple issues and I confirmed these cases could pass with my update.
On top of that, SourceSwap=True is necessary for StoreCInUnroll.
I added code to reject StoreCInUnroll+SourceSwap=False.

Also, I added 2 yaml files under pre_checkin/mfma for zgemm mfma test.

Please check my update.

rosenrodt reviewed (approved):
LGTM
